feat(data_structures)!: put all parts behind features

overlookmotel opened the pull request:
Put all parts of `oxc_data_structures` crate behind features.

`oxc_data_structures` is intended to be a "kitchen sink" crate where any data structures which are shared between multiple crates can go. We want to avoid having tons of little crates `oxc_rope`, `oxc_stack` etc. But this means it'll inevitably grow over time, and that will impact compile times.

Feature-gating each part of the crate should prevent compile times suffering as `oxc_data_structures` grows.

The immediate benefit is it removes the `ropey` dependency from `oxc_ast` and various other crates.

overlookmotel commented:
* **#9849** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9849?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9849?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9849).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Mar 18, 2:41 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
feat(data_structures)!: put all parts behind features (#9849)

Put all parts of `oxc_data_structures` crate behind features.

`oxc_data_structures` is intended to be a "kitchen sink" crate where any data structures which are shared between multiple crates can go. We want to avoid having tons of little crates `oxc_rope`, `oxc_stack` etc. But this means it'll inevitably grow over time, and that will impact compile times.

Feature-gating each part of the crate should prevent compile times suffering as `oxc_data_structures` grows.

The immediate benefit is it removes the `ropey` dependency from `oxc_ast` and various other crates.
